bokehjs/test/unit/models/tools/inspectors/hover_tool.ts

```diff
@@ -0,0 +1,75 @@
+import {expect} from "assertions"
+import {build_view} from "@bokehjs/core/build_views"
+import {ColumnDataSource} from "@bokehjs/models/sources/column_data_source"
+import {Circle, CircleView} from "@bokehjs/models/glyphs/circle"
+import {GlyphRenderer} from "@bokehjs/models/renderers/glyph_renderer"
+import {Plot} from "@bokehjs/models/plots/plot"
+import {Range1d} from "@bokehjs/models/ranges/range1d"
+
+import {HoverTool, HoverToolView} from "@bokehjs/models/tools/inspectors/hover_tool"
+
+export interface HoverToolTestCase {
+  data: {[key: string]: (number | null)[]}
+  data_source: ColumnDataSource
+  hover_tool_view: HoverToolView
+  glyph_view: CircleView
+}
+
+async function make_testcase(): Promise<HoverToolTestCase> {
+  const plot = new Plot({
+    x_range: new Range1d({start: -1, end: 1}),
+    y_range: new Range1d({start: -1, end: 1}),
+  })
+
+  const plot_view = (await build_view(plot)).build()
+
+  const data = {x: [0, 0.5, 1], y: [0, 0.5, 1], z: [null, null, null]}
+  const data_source = new ColumnDataSource({data})
+
+  const glyph = new Circle({
+    x: {field: "x"},
+    y: {field: "y"},
+  })
+
+  const glyph_renderer = new GlyphRenderer({glyph, data_source})
+  const glyph_renderer_view = await build_view(glyph_renderer, {parent: plot_view})
+
+  const hover_tool = new HoverTool({
+    active: true,
+    renderers: [glyph_renderer as any],
+  })
+  plot.add_tools(hover_tool)
+  await plot_view.ready
+
+  const hover_tool_view = plot_view.tool_views.get(hover_tool)! as HoverToolView
+  plot_view.renderer_views.set(glyph_renderer, glyph_renderer_view)
+  return {
+    data,
+    data_source,
+    hover_tool_view,
+    glyph_view: glyph_renderer_view.glyph as CircleView,
+  }
+}
+
+describe("Hover Tool", () => {
+
+  describe("View", () => {
+
+    it("should update internals when changing the tooltips", async () => {
+      const testcase = await make_testcase()
+      testcase.hover_tool_view.initialize()
+      let el        = (testcase.hover_tool_view as any)._template_el as HTMLElement
+      let value_els = el.querySelectorAll<HTMLElement>("[data-value]")
+      expect(value_els.length).to.be.equal(3)
+
+      testcase.hover_tool_view.model.tooltips = [["toto", "$x"]]
+      el        = (testcase.hover_tool_view as any)._template_el as HTMLElement
+      value_els = el.querySelectorAll<HTMLElement>("[data-value]")
```

Comment:
Don't reuse variables. Prefer `const` and enumerated names or split this test case into three.

Reply:
This indentation is unnecessary.

Reply:
We will also need to figure out how to test this without accessing protected/private members and performing `as any` casts. I know there are plenty of unit tests like this in our codebase, but that's an unfortunate historical artifact.

Reply:
Unfortunately almost all the code in the tool is protected. I'm not sure how to proceed without refactoring extensively, something I'm not comfortable with as I'm largely a python coder. 

The simpler solution would be to make one of the methods public. For example `_render_template` seems not to mutate the instance. would that be acceptable ?

 Another option could be to extract the code from `_update` which creates the tooltips without adding them to the html.